src/main/java/com/relayrides/pushy/apns/ApnsClient.java

```diff
@@ -0,0 +1,749 @@
+     *
+     * @return a {@code Future} that will succeed when the client has connected to the gateway and is ready to send
+     * push notifications
+     *
+     * @see ApnsClient#PRODUCTION_APNS_HOST
+     * @see ApnsClient#DEVELOPMENT_APNS_HOST
+     * @see ApnsClient#DEFAULT_APNS_PORT
+     * @see ApnsClient#ALTERNATE_APNS_PORT
+     */
+    public Future<Void> connect(final String host, final int port) {
+        final Future<Void> connectionReadyFuture;
+
+        if (this.bootstrap.group().isShuttingDown() || this.bootstrap.group().isShutdown()) {
+            connectionReadyFuture = new FailedFuture<Void>(GlobalEventExecutor.INSTANCE,
+                    new IllegalStateException("Client's event loop group has been shut down and cannot be restarted."));
+        } else {
+            synchronized (this.bootstrap) {
+                // We only want to begin a connection attempt if one is not already in progress or complete; if we already
+                // have a connection future, just return the existing promise.
+                if (this.connectionReadyPromise == null) {
+                    final ChannelFuture connectFuture = this.bootstrap.connect(host, port);
+                    this.connectionReadyPromise = connectFuture.channel().newPromise();
+
+                    connectFuture.addListener(new GenericFutureListener<ChannelFuture>() {
+
+                        @Override
+                        public void operationComplete(final ChannelFuture future) throws Exception {
+                            if (!future.isSuccess()) {
+                                log.debug("Failed to connect.", future.cause());
+                                ApnsClient.this.connectionReadyPromise.tryFailure(future.cause());
+                            }
+                        }
+                    });
+
+                    connectFuture.channel().closeFuture().addListener(new GenericFutureListener<ChannelFuture> () {
+
+                        @Override
+                        public void operationComplete(final ChannelFuture future) throws Exception {
+                            // We always want to try to fail the "connection ready" promise if the connection closes; if
+                            // it has already succeeded, this will have no effect.
+                            ApnsClient.this.connectionReadyPromise.tryFailure(
+                                    new IllegalStateException("Channel closed before HTTP/2 preface completed."));
+
+                            synchronized (ApnsClient.this.bootstrap) {
+                                ApnsClient.this.connectionReadyPromise = null;
+
+                                if (ApnsClient.this.reconnectionPromise != null) {
+                                    log.debug("Disconnected. Next automatic reconnection attempt in {} seconds.", ApnsClient.this.reconnectDelay);
+
+                                    future.channel().eventLoop().schedule(new Runnable() {
```

Comment:
So my thinking here—and I'm very open to other ideas on this—is that most of the permanent failures (i.e. bogus certificates) would happen on the initial connection attempt. We do NOT automatically retry the initial connection attempt (automatic reconnection attempts only start after we establish the initial connection), and the idea is that we want callers to see the failure and deal with it accordingly. After that, it's more likely that we're dealing with a temporary network outage or a server shutdown or something.

I guess I see this as a choice between two evils:

1. We decline to impose a retry limit on automatic reconnection attempts and (gently!) spin forever if a certificate expires after the initial connection attempt.
2. We *do* impose a retry limit, but guess wrong and wind up permanently disconnecting a client that would have automatically reconnected later.

What do you think?